### ISSUE TITLE
Invalid attaching of the ConsoleResponseSender listener

### ISSUE THREAD

**Issue by fedys** [zendframework/zend-mvc-console#12]
This issue is related to the #10 

Because of the [ConsoleResponseSender](https://github.com/zendframework/zend-mvc-console/blob/1.1.10/src/ResponseSender/ConsoleResponseSender.php) is attached via the [delegator](https://github.com/zendframework/zend-mvc-console/blob/1.1.10/src/Service/ConsoleResponseSenderDelegatorFactory.php#L35) and delegators are called before initializers then the [EventManagerAwareInitializer](https://github.com/zendframework/zend-mvc/blob/release-3.0.2/src/Service/ServiceManagerConfig.php#L105) (re)sets the EventManager which causes lost of all previously attached listeners including the ConsoleResponseSender.

**Comment by kachar** [comment]
Related question: http://stackoverflow.com/questions/38636186/how-to-create-zf3-console-application

**Comment by weierophinney** [comment]
@fedys I think you're mis-diagnosing the root cause. The `EventManagerAwareInitializer` only resets the EM instance _if one is not already present in the instance_ (see the [lines immediately preceding the one you linked](https://github.com/zendframework/zend-mvc/blob/release-3.0.2/src/Service/ServiceManagerConfig.php#L98-L103)).

Note: I'm not saying the problem doesn't exist; I'm saying that the diagnosis is likely incorrect. I'll see if I can resolve it shortly.

**Comment by fedys** [comment]
@weierophinney The problem I described really exists. The root cause is `$eventManager->getSharedManager()` returns null and therefore [the second part of the condition](https://github.com/zendframework/zend-mvc/blob/release-3.0.2/src/Service/ServiceManagerConfig.php#L100) is not met.

**Comment by weierophinney** [comment]
The problem is in zend-mvc's `SendResponseListener` implementation/service configuration.

`SendResponseListener` will lazy-load an event manager instance if none is already present when `getEventManager()` is invoked; however, the instance is created with no arguments, which means that with zend-eventmanager v3, no shared manager is present. Also in v3, there is no way to inject a shared manager after the fact; it must be injected at instantiation (this was a change made to aid performance).

The solution will be to provide a factory for the `SendResponseListener` within zend-mvc that pulls the `EventManager` service and pushes it into a new `SendResponseListener` instance.

You can test this locally in the meantime. Create a factory class as follows, in `module/Application/src/SendResponseListenerFactory.php` (or `module/Application/src/Application/SendResponseListenerFactory.php`, if still using a PSR-0 structure from an older application):

``` php
namespace Application;

use Interop\Container\ContainerInterface;
use Zend\Mvc\SendResponseListener;

class SendResponseListenerFactory
{
    public function __invoke(ContainerInterface $container)
    {
        $listener = new SendResponseListener();
        $listener->setEventManager($container->get('EventManager'));
        return $listener;
    }
}
```

Add the following to `module/Application/config/module.config.php`:

``` php
'service_manager' => [
    'factories' => [
        \Zend\Mvc\SendResponseListener::class => \Application\SendResponseListenerFactory::class,
    ]
]
```

In addition to the above, I tested by adding a single console route, and adding console usage via the `Application\Module::getConsoleUsage()` method. I can verify that all works as expected once the new factory is registered; without the factory registered, no output is generated.

I'll get a PR ready for zend-mvc shortly, and tag this and other related issues when it's ready to merge.

**Comment by weierophinney** [comment]
To get the fix, please update your zend-mvc to 3.0.3:

``` bash
$ composer require "zendframework/zend-mvc:^3.0.3"
```
